### PR TITLE
Move to using ToListAndFree in more places in the parser.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/65702

### CODE REVIEW

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxListPool.cs [hunk 1]
```diff
@@ -55,12 +55,14 @@ internal SyntaxListBuilder Allocate()
...
         {
-            RoslynDebug.Assert(item.UnderlyingBuilder is object);
             Free(item.UnderlyingBuilder);
         }
 
-        internal void Free(SyntaxListBuilder item)
+        internal void Free(SyntaxListBuilder? item)
         {
+            if (item is null)
+                return;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxListPool.cs:64]: simplifies logic for the callers that lazily initialize the builder.

**Comment by CyrusNajmabadi** [comment]
@jjonescz @RikkiGibson ptal :)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -429,7 +429,8 @@ private enum NamespaceParts
                             var attributeLists = _pool.Allocate<AttributeListSyntax>();
                             var modifiers = _pool.Allocate();
 
-                            body.Members.Add(adjustStateAndReportStatementOutOfOrder(ref seen, this.ParseNamespaceDeclaration(attributeLists, modifiers)));
+                            body.Members.Add(adjustStateAndReportStatementOutOfOrder(
+                                ref seen, this.ParseNamespaceDeclaration(attributeLists, modifiers)));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:433]:
```suggestion
                            body.Members.Add(adjustStateAndReportStatementOutOfOrder(ref seen, this.ParseNamespaceDeclaration(attributeLists, modifiers)));
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -3442,11 +3420,7 @@ private ConversionOperatorDeclarationSyntax TryParseConversionOperatorDeclaratio
                     explicitInterfaceOpt = null;
                     opKeyword = this.EatToken(SyntaxKind.OperatorKeyword);
                     type = this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_IdentifierExpected);
-                    SyntaxToken open = SyntaxFactory.MissingToken(SyntaxKind.OpenParenToken);
-                    SyntaxToken close = SyntaxFactory.MissingToken(SyntaxKind.CloseParenToken);
                     var parameters = _pool.AllocateSeparated<ParameterSyntax>();
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:3423]: this is making a builder just for an empty list.  can likely still simplify this further.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -5444,8 +5390,7 @@ private EnumDeclarationSyntax ParseEnumDeclaration(SyntaxList<AttributeListSynta
                 var type = this.ParseType();
                 var tmpList = _pool.AllocateSeparated<BaseTypeSyntax>();
                 tmpList.Add(_syntaxFactory.SimpleBaseType(type));
-                baseList = _syntaxFactory.BaseList(colon, tmpList);
-                _pool.Free(tmpList);
+                baseList = _syntaxFactory.BaseList(colon, _pool.ToListAndFree(tmpList));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:5393]:
```suggestion
                baseList = _syntaxFactory.BaseList(
                    colon,
                    _pool.ToListAndFree(tmpList));
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -5804,13 +5751,14 @@ private SimpleNameSyntax ParseSimpleName(NameOptions options = NameOptions.None)
                 if (kind == ScanTypeArgumentListKind.DefiniteTypeArgumentList || (kind == ScanTypeArgumentListKind.PossibleTypeArgumentList && (options & NameOptions.InTypeList) != 0))
                 {
                     Debug.Assert(this.CurrentToken.Kind == SyntaxKind.LessThanToken);
-                    SyntaxToken open;
+
                     var types = _pool.AllocateSeparated<TypeSyntax>();
-                    SyntaxToken close;
-                    this.ParseTypeArgumentList(out open, types, out close);
+                    this.ParseTypeArgumentList(out var open, types, out var close);
                     name = _syntaxFactory.GenericName(id.Identifier,
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:5757]:
```suggestion
                    name = _syntaxFactory.GenericName(
                        id.Identifier,
```

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -7444,17 +7393,21 @@ private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
...
-                _pool.Free(missingTypes);
-                return funcPtr;
+                return SyntaxFactory.FunctionPointerType(
+                    @delegate,
+                    asterisk,
+                    callingConvention,
+                    SyntaxFactory.FunctionPointerParameterList(
+                        lessThanTokenError,
+                        _pool.ToListAndFree(missingTypes),
+                        TryEatToken(SyntaxKind.GreaterThanToken) ?? SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken)));
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7410]: fwiw, this pattern is all over the place in the parser. :)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -13080,27 +13010,25 @@ private ParameterSyntax ParseLambdaParameter()
...
-            var parameter = _syntaxFactory.Parameter(attributes, modifiers.ToList(), paramType, identifier, @default: equalsValueClause);
-            _pool.Free(modifiers);
-            return parameter;
+            return _syntaxFactory.Parameter(
+                attributes,
+                _pool.ToTokenListAndFree(modifiers),
+                paramType,
+                identifier,
+                equalsToken != null
+                    ? _syntaxFactory.EqualsValueClause(equalsToken, this.ParseExpressionCore())
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13030]: fwiw, this pattern is all over the place in the parser. :)

**Review by jjonescz — APPROVED** [review]

**Pull request merged**